src/main/java/com/linkedin/kafka/clients/utils/DefaultHeaderSerde.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2017 LinkedIn Corp. Licensed under the BSD 2-Clause License (the "License").
+ * See License in the project root for license information.
+ */
+package com.linkedin.kafka.clients.utils;
+
+import java.nio.ByteBuffer;
+
+
+/**
+ * Utilities to implement header serializer and deserializer.
+ */
+public interface DefaultHeaderSerde {
+  /**
+   *   All this bit manipulation makes this number look like an invalid UTF8 encoded string if someone starts to read it
+   * within the first 7 bytes
+   */
+  long DEFAULT_HEADER_MAGIC
+    = (0x4c6d4eef4b7a44L | 0b11000000_11000000_11000000_11000000_11000000_11000000_11000000_11000000L) &
+    0b11011111_11011111_11011111_11011111_11011111_11011111_11011111_11011111L;
+
+  int MAGIC_SIZE = 8;
+
+  byte VERSION_1 = 1;
+
+  int VERSION_AND_FLAGS_SIZE = 1;
+
+  int ALL_HEADER_SIZE_FIELD_SIZE = 4;
+
+  byte USER_VALUE_IS_NULL_FLAG = 0x10;
+
+  /**
+   * The size of the size field that encoded the key length.
+   */
+  int KEY_SIZE_SIZE = 1;
+
+  /**
+   * The size of the value size field.
+   */
+  int VALUE_SIZE_SIZE = 4;
+
+  /**
+   * @return DEFAULT_HEADER_MAGIC as byte array.
+   */
+  static byte[] defaultHeaderMagicBytes() {
+    ByteBuffer bbuf = ByteBuffer.allocate(MAGIC_SIZE);
+    bbuf.putLong(DEFAULT_HEADER_MAGIC);
+    return bbuf.array();
+  }
+
+  /**
+   * The length of a utf8 encoded string.
+   * @param s non-null
+   * @return a non-negative integer. The number of bytes in the utf-8 encoded form of the string
+   */
+  static int utf8StringLength(String s) {
```

Comment:
I'm trying to avoid serializing the string twice: once to get the size and the second time to serialize it into byte buffer.